File: antd-py/src/antd/_discover.py
"""Daemon port-file discovery for antd.

The antd daemon writes a ``daemon.port`` file on startup containing:
  - Line 1: REST port
  - Line 2: gRPC port

This module reads that file to auto-discover the daemon's listen addresses.
"""

from __future__ import annotations

import os
import sys
from pathlib import Path

_PORT_FILE_NAME = "daemon.port"
_DATA_DIR_NAME = "ant"


def discover_daemon_url() -> str:
    """Return the REST base URL from the daemon port file, or ``""`` on failure."""
    rest, _ = _read_port_file()
    if rest == 0:
        return ""
    return f"http://127.0.0.1:{rest}"


def discover_grpc_target() -> str:
    """Return the gRPC target from the daemon port file, or ``""`` on failure."""
    _, grpc = _read_port_file()
    if grpc == 0:
        return ""
    return f"127.0.0.1:{grpc}"


def _read_port_file() -> tuple[int, int]:
    """Read the daemon.port file and return ``(rest_port, grpc_port)``.

    A single-line file is valid (gRPC port will be 0).
    Returns ``(0, 0)`` on any error.
    """
    dir_path = _data_dir()
    if not dir_path:
        return 0, 0

    port_file = Path(dir_path) / _PORT_FILE_NAME
    try:
        text = port_file.read_text(encoding="utf-8")
    except OSError:
        return 0, 0

    lines = text.strip().splitlines()
    if not lines:
        return 0, 0

    rest_port = _parse_port(lines[0])
    grpc_port = _parse_port(lines[1]) if len(lines) >= 2 else 0
    return rest_port, grpc_port


def _parse_port(s: str) -> int:
    """Parse a port string, returning 0 on failure."""
    try:
        n = int(s.strip())
    except (ValueError, TypeError):
        return 0
    if 1 <= n <= 65535:
        return n
    return 0


def _data_dir() -> str:
    """Return the platform-specific data directory for ant, or ``""``."""
    if sys.platform == "win32":
        appdata = os.environ.get("APPDATA", "")
        if not appdata:
            return ""
        return os.path.join(appdata, _DATA_DIR_NAME)

    if sys.platform == "darwin":
        home = os.environ.get("HOME", "")
        if not home:
            return ""
        return os.path.join(home, "Library", "Application Support", _DATA_DIR_NAME)

    # Linux and others
    xdg = os.environ.get("XDG_DATA_HOME", "")
    if xdg:
        return os.path.join(xdg, _DATA_DIR_NAME)
    home = os.environ.get("HOME", "")
    if not home:
        return ""
    return os.path.join(home, ".local", "share", _DATA_DIR_NAME)

File: antd-cpp/src/discover.cpp
#include "antd/discover.hpp"

#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <string>

namespace fs = std::filesystem;

namespace antd {
namespace {

constexpr const char* kPortFileName = "daemon.port";
constexpr const char* kDataDirName = "ant";

/// Parse a port number (1-65535) from a string. Returns 0 on failure.
uint16_t parse_port(const std::string& s) {
    try {
        unsigned long n = std::stoul(s);
        if (n == 0 || n > 65535) return 0;
        return static_cast<uint16_t>(n);
    } catch (...) {
        return 0;
    }
}

/// Return the platform-specific data directory for ant.
///   - Windows: %APPDATA%\ant
///   - macOS:   ~/Library/Application Support/ant
///   - Linux:   $XDG_DATA_HOME/ant or ~/.local/share/ant
fs::path data_dir() {
#ifdef _WIN32
    const char* appdata = std::getenv("APPDATA");
    if (!appdata || appdata[0] == '\0') return {};
    return fs::path(appdata) / kDataDirName;
#elif defined(__APPLE__)
    const char* home = std::getenv("HOME");
    if (!home || home[0] == '\0') return {};
    return fs::path(home) / "Library" / "Application Support" / kDataDirName;
#else
    const char* xdg = std::getenv("XDG_DATA_HOME");
    if (xdg && xdg[0] != '\0') {
        return fs::path(xdg) / kDataDirName;
    }
    const char* home = std::getenv("HOME");
    if (!home || home[0] == '\0') return {};
    return fs::path(home) / ".local" / "share" / kDataDirName;
#endif
}

/// Read the daemon.port file and return the REST and gRPC ports.
/// The file format is two lines: REST port on line 1, gRPC port on line 2.
std::pair<uint16_t, uint16_t> read_port_file() {
    auto dir = data_dir();
    if (dir.empty()) return {0, 0};

    auto path = dir / kPortFileName;
    std::ifstream ifs(path);
    if (!ifs.is_open()) return {0, 0};

    std::string line1, line2;
    if (!std::getline(ifs, line1)) return {0, 0};
    std::getline(ifs, line2);  // optional second line

    return {parse_port(line1), parse_port(line2)};
}

}  // namespace

std::string discover_daemon_url() {
    auto [rest, grpc] = read_port_file();
    (void)grpc;
    if (rest == 0) return {};
    return "http://127.0.0.1:" + std::to_string(rest);
}

std::string discover_grpc_target() {
    auto [rest, grpc] = read_port_file();
    (void)rest;
    if (grpc == 0) return {};
    return "127.0.0.1:" + std::to_string(grpc);
}

}  // namespace antd
